src/core/solvers/ipopt/ipopt-iface.cpp

```diff
@@ -0,0 +1,665 @@
+    datas_[t]->u = Eigen::VectorXd::Map(x + ixu + ndxi, nui);
+
+    model->get_state()->integrate(xs_[t], datas_[t]->dx, datas_[t]->x);
+    model->get_state()->Jintegrate(xs_[t], datas_[t]->dx, datas_[t]->Jsum_dx, datas_[t]->Jsum_dx, second, setto);
+
+    model->calc(data, datas_[t]->x, datas_[t]->u);
+    model->calcDiff(data, datas_[t]->x, datas_[t]->u);
+    datas_[t]->Ldx = datas_[t]->Jsum_dx.transpose() * data->Lx;
+
+    for (std::size_t j = 0; j < ndxi; j++) {
+      grad_f[ixu + j] = datas_[t]->Ldx(j);
+    }
+
+    for (std::size_t j = 0; j < nui; j++) {
+      grad_f[ixu + ndxi + j] = data->Lu(j);
+    }
+    ixu += ndxi + nui;
+  }
+
+  {
+    // Terminal model
+    const boost::shared_ptr<ActionModelAbstract> &model = problem_->get_terminalModel();
+    const boost::shared_ptr<ActionDataAbstract> &data = problem_->get_terminalData();
+    const std::size_t ndxi = model->get_state()->get_ndx();
+
+    datas_[T_]->dx = Eigen::VectorXd::Map(x + ixu, ndxi);
+
+    model->get_state()->integrate(xs_[T_], datas_[T_]->dx, datas_[T_]->x);
+    model->get_state()->Jintegrate(xs_[T_], datas_[T_]->dx, datas_[T_]->Jsum_dx, datas_[T_]->Jsum_dx, second, setto);
+
+    model->calc(data, datas_[T_]->x);
+    model->calcDiff(data, datas_[T_]->x);
+    datas_[T_]->Ldx.noalias() = datas_[T_]->Jsum_dx.transpose() * data->Lx;
+
+    for (std::size_t j = 0; j < ndxi; j++) {
+      grad_f[ixu + j] = datas_[T_]->Ldx(j);
+    }
+  }
+
+  return true;
+}
+
+bool IpoptInterface::eval_g(Ipopt::Index n, const Ipopt::Number *x, bool new_x, Ipopt::Index m, Ipopt::Number *g) {
+  assert_pretty(n == nvar_, "Inconsistent number of decision variables");
+  assert_pretty(m == nconst_, "Inconsistent number of constraints");
+
+  std::size_t ixu = 0;
+  std::size_t ix = 0;
+  // Dynamic constraints
+  for (std::size_t t = 0; t < T_; t++) {
+    const boost::shared_ptr<ActionModelAbstract> &model = problem_->get_runningModels()[t];
+    const boost::shared_ptr<ActionDataAbstract> &data = problem_->get_runningDatas()[t];
+    const std::size_t ndxi = model->get_state()->get_ndx();
+    const std::size_t nui = model->get_nu();
+
+    const boost::shared_ptr<ActionModelAbstract> &model_next =
+        t + 1 == T_ ? problem_->get_terminalModel() : problem_->get_runningModels()[t + 1];
+    const boost::shared_ptr<ActionDataAbstract> &data_next =
+        t + 1 == T_ ? problem_->get_terminalData() : problem_->get_runningDatas()[t + 1];
+    const std::size_t ndxi_next = model_next->get_state()->get_ndx();
+
+    datas_[t]->dx = Eigen::VectorXd::Map(x + ixu, ndxi);
+    datas_[t]->u = Eigen::VectorXd::Map(x + ixu + ndxi, nui);
+    datas_[t + 1]->dx = Eigen::VectorXd::Map(x + ixu + ndxi + nui, ndxi_next);
+
+    model->get_state()->integrate(xs_[t], datas_[t]->dx, datas_[t]->x);
+    model_next->get_state()->integrate(xs_[t + 1], datas_[t + 1]->dx, datas_[t + 1]->x);
+
+    model->calc(data, datas_[t]->x, datas_[t]->u);
```

Comment:
I do believe that Ipopt should call always eval_f or eval_g first. If so, it means that we need to call `calc` in one of these functions.

Can you check this?

Reply:
calc() is called first at `eval_grad_f`. However, I've only tested this with the few examples I have.

If we want to optimize properly the calls to `calc` and `calcDiff` I think we should do it by considering the IPOPT algorithm. Otherwise, we might be skipping some case.

I'd take a safe path in this case and leave the calls as they are. Unfortunately, at this moment, I do not have the time to do such optimization.